src/coreclr/tools/Common/Compiler/ObjectWriter/WasmObjectWriter.cs

```diff
@@ -0,0 +1,246 @@
+
+        public override void EmitObject(Stream outputFileStream, IReadOnlyCollection<DependencyNode> nodes, IObjectDumper dumper, Logger logger)
+        {
+            ArrayBuilder<WasmFuncType> methodSignatures = new();
+            ArrayBuilder<IMethodBodyNode> methodBodies = new();
+            foreach (DependencyNode node in nodes)
+            {
+                if (node is not ObjectNode)
+                {
+                    continue;
+                }
+
+                if (node is IMethodBodyNode methodNode)
+                {
+                    methodSignatures.Add(_wasmAbiContext.GetSignature(methodNode.Method));
+                    methodBodies.Add(methodNode);
+                    // TODO: record relocations and attached data
+                }
+            }
+
+            Dictionary<WasmFuncType, int> signatureMap = new();
+            int signatureIndex = 0;
+            for (int i = 0; i < methodSignatures.Count; i++)
+            {
+                if (!signatureMap.ContainsKey(methodSignatures[i]))
+                {
+                    signatureMap[methodSignatures[i]] = signatureIndex++;
+                }
+            }
+
+            WasmFuncType[] uniqueSignatures = new WasmFuncType[signatureMap.Count];
+            foreach (var kvp in signatureMap)
+            {
+                uniqueSignatures[kvp.Value] = kvp.Key;
+            }
+
+            EmitWasmHeader(outputFileStream);
+            EmitSection(() => WriteTypeSection(uniqueSignatures, logger), outputFileStream, logger);
+            EmitSection(() => WriteFunctionSection(methodSignatures.ToArray(), signatureMap, logger), outputFileStream, logger);
+            EmitSection(() => WriteExportSection(methodBodies.ToArray(), methodSignatures.ToArray(), signatureMap, logger), outputFileStream, logger);
+            EmitSection(() => WriteCodeSection(methodBodies.ToArray(), logger),  outputFileStream, logger);
+        }
+
+        private WasmSection WriteExportSection(IReadOnlyCollection<IMethodBodyNode> methodNodes, IReadOnlyCollection<WasmFuncType> methodSignatures,
+            Dictionary<WasmFuncType, int> signatureMap, Logger logger)
+        {
+            WasmSection exportSection = new WasmSection(WasmSectionType.Export, new SectionData(), "export");
+            SectionWriter writer = exportSection.Writer;
+            // Write the number of exports
+            writer.WriteULEB128((ulong)methodNodes.Count);
+            for (int i = 0; i < methodNodes.Count; i++)
+            {
+                var methodNode = methodNodes.ElementAt(i);
+                var methodSignature = methodSignatures.ElementAt(i);
+                string exportName = methodNode.GetMangledName(_nodeFactory.NameMangler);
+
+                var length = Encoding.UTF8.GetByteCount(exportName);
+                writer.WriteULEB128((ulong)length);
+                writer.WriteUtf8StringNoNull(exportName);
+                writer.WriteByte(0x00); // export kind: function
```

Comment:
Nit: We may want an enum or named constants for this since we'll want to export non-functions eventually, but this is fine for now if you don't want to change it. Then it's self-explanatory